File: server/src/main/java/br/com/fastgondolas/fastos/server/model/OrdemDeServico.java
package br.com.fastgondolas.fastos.server.model;

import br.com.fastgondolas.fastos.server.dto.CategoriaDto;
import br.com.fastgondolas.fastos.server.dto.PecaPendenteDto;
import br.com.fastgondolas.fastos.server.model.OsStatus;
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.Type;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "ordens_de_servico")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class OrdemDeServico extends BaseEntity{

    @Column(name = "numero_os")
    private String numeroOs;

    private String numeroSerieEquipamento;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "problema_detalhado", columnDefinition = "jsonb")
    private List<CategoriaDto> problemaDetalhado;

    private String diagnosticoTecnico;
    private String solucaoAplicada;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "status")
    private OsStatus status;

    private String motivoPausa;

    @Type(JsonBinaryType.class)
    @Column(columnDefinition = "jsonb")
    private List<PecaPendenteDto> pecasPendentes;

    private String descricaoPendencia;
    private Instant dataAgendamento;
    private Instant dataInicioExecucao;
    private Instant dataFimExecucao;
    private String codigoClienteExterno;
    private String numeroPedidoExterno;
    private LocalDate dataFaturamento;
    private boolean emGarantia;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cliente_id")
    private Cliente cliente;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "endereco_id")
    private Endereco endereco;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "tecnico_responsavel_id")
    private User tecnicoResponsavel;

    @OneToMany(mappedBy = "ordemDeServico", fetch = FetchType.LAZY)
    private List<Anexo> anexos;

    @ManyToMany(mappedBy = "ordensDeServico", fetch = FetchType.LAZY)
    private List<Visita> visitas;
}

File: mobile/src/db/models/OrdemDeServico.ts
import { Model, Relation } from '@nozbe/watermelondb';
import {
  date,
  field,
  json,
  readonly,
  relation,
  text,
} from '@nozbe/watermelondb/decorators';
import Cliente from './Cliente';
import Endereco from './Endereco';
import Tecnico from './Tecnico';
import Visita from './Visita';
import type { DefectFormState } from '../../types/defect';

const sanitizeJson = (rawJson: any) => rawJson;

export default class OrdemDeServico extends Model {
  static override table = 'ordens_de_servico';

  static override associations = {
    clientes: { type: 'belongs_to', key: 'cliente_id' },
    enderecos: { type: 'belongs_to', key: 'endereco_id' },
    tecnicos: { type: 'belongs_to', key: 'tecnico_responsavel_id' },
    visitas: { type: 'belongs_to', key: 'visita_id' },
  } as const;

  @text('numero_os') numeroOs!: string;
  @field('status') status!: string;

  @json('problema_detalhado', sanitizeJson) problemaDetalhado!: DefectFormState;
  @text('diagnostico_tecnico') diagnosticoTecnico!: string;
  @text('solucao_aplicada') solucaoAplicada!: string;

  @field('cliente_id') clienteId!: string;
  @field('endereco_id') enderecoId!: string;
  @field('tecnico_responsavel_id') tecnicoResponsavelId!: string;
  @field('visita_id') visitaId!: string;

  @date('data_inicio_execucao') dataInicioExecucao?: Date;
  @date('data_fim_execucao') dataFimExecucao?: Date;

  @relation('clientes', 'cliente_id') cliente!: Relation<Cliente>;
  @relation('enderecos', 'endereco_id') endereco!: Relation<Endereco>;
  @relation('tecnicos', 'tecnico_responsavel_id')
  tecnicoResponsavel!: Relation<Tecnico>;
  @relation('visitas', 'visita_id') visita!: Relation<Visita>;

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;
}
